framework/http/FileStream.php

```diff
@@ -0,0 +1,263 @@
+<?php
+/**
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+
+namespace yii\http;
+
+use Psr\Http\Message\StreamInterface;
+use Yii;
+use yii\base\BaseObject;
+use yii\base\ErrorHandler;
+use yii\base\InvalidConfigException;
+
+/**
+ * FileStream represents file stream.
+ *
+ * Example:
+ *
+ * ```php
+ * $stream = new FileSteam([
+ *     'filename' => '@app/files/items.txt',
+ *     'mode' => 'w+',
+ * ]);
+ *
+ * $stream->write('some content...');
+ * $stream->close();
+ * ```
+ *
+ * @author Paul Klimov <klimov.paul@gmail.com>
+ * @since 2.1.0
+ */
+class FileStream extends BaseObject implements StreamInterface
+{
+    /**
+     * @var string file or stream name.
+     * Path alias can be used here, for example: '@app/runtime/items.csv'.
+     * This field can also be PHP stream name, e.g. anything which can be passed to `fopen()`, for example: 'php://input'.
+     */
+    public $filename;
```

Comment:
File name seems to be mandatory. Have you thought about making it a constructor argument?

Reply:
I don't think that it's a good reason to affect class design. If application can't configure pretty much regular PHP class with constructor then we need to address it in the application.

Reply:
Mode is, as well, mandatory so it should be constructor argument.

Reply:
Another argument for making these constructor arguments and private is that changing these values after opening a stream makes no sense.

Reply:
> I don't think that it's a good reason to affect class design. If application can't configure pretty much regular PHP class with constructor then we need to address it in the application.

I would like to keep a consistentcy comparing other Yii classes, which in most cases do not introduce constructor parameters.
Also it is actually unlikely developer will create these stream objects anywhere manually.
For the file response sending he will likely use `Response::sendFile()`.
I do not think it is a crucial.

> argument for making these constructor arguments and private is that changing these values after 
opening a stream makes no sense.

Same can be said for example for [yii\db\Connection::$dsn](https://github.com/yiisoft/yii2/blob/master/framework/db/Connection.php#L164): after connection opened changing of the `$dsn`, `$username` or `$password` make no effect and no sense.

Reply:
> I do not think it is a crucial.

It is not but:

1. It's not good to develop this way with mandatory dependencies set via public properties. Makes too much room for errors.
2. It serves as an example and it's a bad example. I know we have that basically everywhere but we have to start with something.
3. It's a good test for our DI container capabilities.

> Same can be said for example for yii\db\Connection::$dsn: after connection opened changing of the $dsn, $username or $password make no effect and no sense.

Yes. *Ideally* it should be changed as well...